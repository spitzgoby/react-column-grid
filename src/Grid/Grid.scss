@import '../style/sizes.scss';

.rcg-c {
  display: grid;
  grid-template-columns: repeat(12, 1fr);
}

@media screen and (min-width: 1535px) {
  .rcg-xl_h {
    display: none;
  }
}

@each $sizeName, $maxWidth in $maxWidthSizes {
  @media screen and (max-width: #{$maxWidth}) {
    .rcg-#{$sizeName}_h {
      display: none;
    }
  }
}

@each $sizeName, $minWidth in $minWidthSizes {
  @media screen and (min-width: #{$minWidth}) {
    @for $col from 1 through 12 {
      @for $width from 1 through 12 {
        @if ($col + $width <= 13) {
          .rcg-#{$sizeName}-#{$col}-#{$width} {
            grid-column: #{$col} / span #{$width};
          }
        }
      }
    }
  }
}