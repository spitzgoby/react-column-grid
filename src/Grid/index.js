import GridComponent from "./Grid";

export const Grid = GridComponent;
export default GridComponent;
