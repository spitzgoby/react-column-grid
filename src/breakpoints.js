export const breakpoints = [
    {
        size: "xs",
        maxWidth: "599px",
    },
    {
        size: "sm",
        maxWidth: "899px",
        minWidth: "600px",
    },
    {
        size: "md",
        maxWidth: "1199px",
        minWidth: "900px",
    },
    {
        size: "lg",
        maxWidth: "1535px",
        minWidth: "1200px",
    },
    {
        size: "xl",
        minWidth: "1536px",
    },
];
export const sizes = breakpoints.map((breakpoint) => breakpoint.size);

export const addMissingSizes = (propName, prop = {}, defaultValue, shorthand) =>
    sizes.reduce((acc, size, index) => {
        if (!Object.prototype.hasOwnProperty.call(prop, size)) {
            if (index > 0) {
                acc[size] = acc[sizes[index - 1]];
            } else if (shorthand(propName, prop)) {
                acc[size] = prop;
            } else {
                acc[size] = defaultValue;
            }
        } else {
            acc[size] = prop[size];
        }

        return acc;
    }, {});

export const createScreenMediaQuery = (breakpoint) =>
    "@media screen" +
    `${breakpoint.minWidth ? ` and (min-width: ${breakpoint.minWidth})` : ""}` +
    `${breakpoint.maxWidth ? ` and (max-width: ${breakpoint.maxWidth})` : ""}`;
