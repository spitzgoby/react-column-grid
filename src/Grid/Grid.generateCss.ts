import { BreakpointsDefinition } from "../utils/breakpoints";
import { generateScreenQueries } from "../utils/mediaQueries";
import { sizes } from "../utils/breakpoints";

const containerDisplay = 'display: grid;';
export const generateGridContainerCss = (columns: number) => {
    const containerTemplateColumns = `grid-template-columns: repeat(${columns}, 1fr);`;

    return `.rcg-c{${containerDisplay}${containerTemplateColumns}}`;
}

export const generateGridBreakpointCss = (breakpoints: BreakpointsDefinition, columns: number) => {
    let columnSpanClassDefinition: string[] = sizes.map((size) => {
        let sizeDefintions: string[] = [`.rcg-${size}_h{display: none;}`];        

        for (let col = 1; col <= columns; col++) {
            for (let width = 1; width <= columns; width++) {
                if (col + width <= columns + 1) {
                    sizeDefintions.push(`.rcg-${size}-${col}-${width}{grid-column:${col}/ span ${width};}`) ;
                } else {
                    width = columns;
                }
            }
        }

        return sizeDefintions.join('');
    });

    return generateScreenQueries(breakpoints, columnSpanClassDefinition);
}