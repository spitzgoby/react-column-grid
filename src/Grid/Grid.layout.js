import { sizes } from "../utils/breakpoints";
import { DEFAULT_GAP } from "../utils/gap";

export const getGap = (gap) =>
    gap || gap === 0
        ? !isNaN(gap) && !isNaN(parseFloat(gap))
            ? gap + "em"
            : gap
        : DEFAULT_GAP;

export const areValidColumns = (colWidth, colOffset, numColumns) =>
    !isNaN(parseInt(colWidth, 10)) &&
    !isNaN(parseInt(colOffset, 10)) &&
    colWidth > 0 &&
    colOffset + colWidth <= numColumns;

const NUM_COLUMNS = 12;
export const getAdjustedLayoutProps = (
    size,
    offset,
    width,
    hide,
    clear,
    columns
) => {
    const sizeHide = hide?.[size];
    const sizeOffset = parseInt(offset?.[size], 10) || 0;
    const sizeWidth = parseInt(width?.[size], 10);
    const suggestedOffset = columns[size] + sizeOffset;
    let adjustedColumn = columns[size];
    let adjustedOffset = sizeHide ? 0 : sizeOffset;

    if (areValidColumns(sizeWidth, suggestedOffset, NUM_COLUMNS) && !sizeHide) {
        adjustedOffset = suggestedOffset;
        adjustedColumn = !clear[size] ? suggestedOffset + sizeWidth : 0;
    } else if (
        areValidColumns(sizeWidth, sizeOffset, NUM_COLUMNS) &&
        !sizeHide
    ) {
        adjustedOffset = sizeOffset;
        adjustedColumn = !clear[size] ? sizeOffset + sizeWidth : 0;
    }

    return {
        adjustedOffset,
        adjustedColumn,
    };
};

export const generateSizeClassNames = ({ hide, item, offset, width }) => {
    return sizes.reduce((acc, size) => {
        acc[`rcg-${size}_h`] = hide[size];
        acc[`rcg-${size}-${offset[size] + 1}-${width[size]}`] = item;

        return acc;
    }, {});
};
