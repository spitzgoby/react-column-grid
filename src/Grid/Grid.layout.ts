import { Breakpoint, sizes } from "../utils/breakpoints";
import { DEFAULT_GAP } from "../constants/gap";
import { 
    getValueOfNumeric, 
    Numeric, 
    numericIsDecimal,
    numericIsInteger 
} from "../utils/numeric";

export type NumericBreakpointValues = {
    xs?: Numeric,
    sm?: Numeric,
    md?: Numeric,
    lg?: Numeric,
    xl?: Numeric
};
export type BooleanBreakpointValues = {
    xs?: boolean,
    sm?: boolean,
    md?: boolean,
    lg?: boolean,
    xl?: boolean
}

const gapHasNoUnits = (gap: string) => numericIsInteger(gap) || numericIsDecimal(gap);
export const getGap = (gap: Numeric): string => {
    const gapType = typeof gap;
    let result: string = DEFAULT_GAP;

    if (gapType === 'string') {
        if (gapHasNoUnits(gap as string)) {
            result = gap + 'em';
        } else {
            result = gap as string 
        }
    } else if (gapType === 'number') {
        result = gap + 'em'; 
    }     
    
    return result;
}

const NUM_COLUMNS = 12;
const areValidColumns = (colWidth: Numeric, colOffset: Numeric): boolean => {
    const width = getValueOfNumeric(colWidth);
    const offset = getValueOfNumeric(colOffset);

    return !isNaN(width) &&
    !isNaN(offset) &&
    width > 0 &&
    width + offset <= NUM_COLUMNS;
}

export const getAdjustedLayoutProps = (
    size: Breakpoint,
    offset: NumericBreakpointValues,
    width: NumericBreakpointValues,
    hide: BooleanBreakpointValues,
    clear: BooleanBreakpointValues,
    columns: NumericBreakpointValues
) => {
    const sizeHide = hide?.[size];
    const sizeOffset = getValueOfNumeric(offset?.[size]) || 0;
    const sizeWidth = getValueOfNumeric(width?.[size]);
    const suggestedOffset = getValueOfNumeric(columns[size]) + sizeOffset;
    let adjustedColumn = columns[size];
    let adjustedOffset = sizeHide ? 0 : sizeOffset;

    if (areValidColumns(sizeWidth, suggestedOffset) && !sizeHide) {
        adjustedOffset = suggestedOffset;
        adjustedColumn = !clear[size] ? suggestedOffset + sizeWidth : 0;
    } else if (
        areValidColumns(sizeWidth, sizeOffset) &&
        !sizeHide
    ) {
        adjustedOffset = sizeOffset;
        adjustedColumn = !clear[size] ? sizeOffset + sizeWidth : 0;
    }

    return {
        adjustedOffset,
        adjustedColumn,
    };
};

type SizeClassNamesParams = { 
    hide: BooleanBreakpointValues, 
    item: boolean, 
    offset: NumericBreakpointValues, 
    width: NumericBreakpointValues
};
type SizeClassNames = {
    [key: string]: boolean,
}
export const generateSizeClassNames = ({ hide, item, offset, width }: SizeClassNamesParams) => {
    return sizes.reduce((acc: SizeClassNames, size: Breakpoint) => {
        acc[`rcg-${size}_h`] = hide[size];
        acc[`rcg-${size}-${getValueOfNumeric(offset[size]) + 1}-${width[size]}`] = item;

        return acc;
    }, {});
};
