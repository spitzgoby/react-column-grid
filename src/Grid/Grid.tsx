import { 
    getValueOfNumeric, 
    Numeric, 
    numericIsInteger 
} from '../utils/numeric';
import classNames from "classnames";
import {
    BooleanBreakpointValues,
    generateContainerClassName,
    generateSizeClassNames,
    getAdjustedLayoutProps,
    NumericBreakpointValues
} from "./Grid.layout";
import PropTypes, { ReactNodeLike } from "prop-types";
import React, { useContext } from "react";
import { GridProvider, GridContext } from '../GridProvider';
import { addMissingSizes, DEFAULT_SCREEN_WIDTHS, ScreenWidths, sizes } from "../utils/breakpoints";

import { DEFAULT_GAP } from "../utils/gap";
import { DEFAULT_COLUMNS, INITIAL_ID } from '../GridProvider/GridContext';

const defaultClear = false;
const defaultHide = false;
const defaultOffset = 0;
const defaultWidth = 12;
const booleanShortHandProps = ["clear"];
const numericShortHandProps = ["offset", "width"];

const useShorthandSyntax = (propName: string, prop: Numeric | boolean): boolean => {
    let result = false;    

    if (typeof prop === 'boolean') { 
        result = booleanShortHandProps.includes(propName);
    } else if (numericShortHandProps.includes(propName)) {
        result = numericIsInteger(prop);
    }

    return result;
}

type Props = {
    breakpoints?: ScreenWidths,
    children: ReactNodeLike,
    className?: string,
    clear?: boolean | BooleanBreakpointValues,
    columns?: number,
    container?: boolean,
    gap?: Numeric,
    hide?: BooleanBreakpointValues,
    item?: boolean,
    offset?: NumericBreakpointValues | Numeric,
    width?: NumericBreakpointValues | Numeric,
}

const Grid: React.FC<Props> = (props) => {
    const {
        breakpoints: propBreakpoints,
        children,
        className,
        columns: propColumns,
        container,
        gap: propGap,
        hide = {},
        item,
        offset = {},
        width = { xs: 12 },
    } = props;
    const { 
        breakpoints: contextBreakpoints, 
        columns: contextColumns, 
        gap: contextGap,
        id 
    } = useContext(GridContext);
    const breakpoints = propBreakpoints || contextBreakpoints || DEFAULT_SCREEN_WIDTHS;
    const columns = propColumns || contextColumns || DEFAULT_COLUMNS;
    const gap = propGap || contextGap || DEFAULT_GAP;    

    // Fill out incomplete layout props
    const adjustedHide = addMissingSizes(
        "hide",
        hide,
        defaultHide,
        useShorthandSyntax
    );
    const adjustedOffset = addMissingSizes(
        "offset",
        offset,
        defaultOffset,
        useShorthandSyntax
    );
    const adjustedWidth = addMissingSizes(
        "width",
        width,
        defaultWidth,
        useShorthandSyntax
    );

    const hasPropsThatOverrideContext = () => 
        breakpoints !== contextBreakpoints
        || columns !== contextColumns
        || gap !== contextGap;
    const shouldRenderWithProvider = () => container && (id === INITIAL_ID || hasPropsThatOverrideContext());

    // Generate classes based on width, offset and sizes
    const sizeClassNames = generateSizeClassNames({
        hide: adjustedHide,
        item,
        offset: adjustedOffset,
        width: adjustedWidth,
    }, id);

    const getClass = () =>
        classNames({ [generateContainerClassName(id)]: container }, sizeClassNames, className);

    const renderChildren = () => {
        let currentColumns = { xs: 0, sm: 0, md: 0, lg: 0, xl: 0 };
        const childrenToRender = Array.isArray(children)
            ? children
            : [children];

        return childrenToRender.map((child, index) => {
            let renderedChild = child;

            if (child?.type?.name === Grid.name) {
                let renderedChildProps = { 
                    ...child.props,
                    key: index
                };

                if (child?.props?.item) {
                    const renderedChildOffset = sizes.reduce((acc: NumericBreakpointValues, size) => {
                        const completedClear = addMissingSizes(
                            "clear",
                            child.props.clear,
                            defaultClear,
                            useShorthandSyntax
                        );
                        const completedHide = addMissingSizes(
                            "hide",
                            child.props.hide,
                            defaultHide,
                            useShorthandSyntax
                        );
                        const completedOffset = addMissingSizes(
                            "offset",
                            child.props.offset,
                            defaultOffset,
                            useShorthandSyntax
                        );
                        const completedWidth = addMissingSizes(
                            "width",
                            child.props.width,
                            defaultWidth,
                            useShorthandSyntax
                        );
                        const { adjustedColumn, adjustedOffset } =
                            getAdjustedLayoutProps(
                                size,
                                completedOffset,
                                completedWidth,
                                completedHide,
                                completedClear,
                                currentColumns,
                                columns
                            );

                        currentColumns[size] = getValueOfNumeric(adjustedColumn);
                        acc[size] = getValueOfNumeric(adjustedOffset);

                        return acc;
                    }, {});

                    renderedChildProps = {
                        ...renderedChildProps,
                        offset: renderedChildOffset,
                    };
                } 

                renderedChild = <Grid {...renderedChildProps} />;
            }

            return renderedChild;
        });
    };

    const renderWithoutGridProvider = () => (
        <div className={getClass()}>
            <>{container ? renderChildren() : children}</>
        </div>
    );

    const renderWithGridProvider = () => (
        <GridProvider breakpoints={breakpoints} columns={columns} gap={gap}>
            <Grid {...props} />
        </GridProvider>
    )

    return (
        shouldRenderWithProvider() ? renderWithGridProvider() : renderWithoutGridProvider()
    );
};

Grid.propTypes = {
    breakpoints: PropTypes.arrayOf(PropTypes.number),
    children: PropTypes.node,
    clear: PropTypes.oneOfType([
        PropTypes.bool, 
        PropTypes.shape({
            xs: PropTypes.bool,
            sm: PropTypes.bool,
            md: PropTypes.bool,
            lg: PropTypes.bool,
            xl: PropTypes.bool,
        })
    ]),
    className: PropTypes.string,
    columns: PropTypes.number,
    container: PropTypes.bool,
    gap: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    hide: PropTypes.shape({
        xs: PropTypes.bool,
        sm: PropTypes.bool,
        md: PropTypes.bool,
        lg: PropTypes.bool,
        xl: PropTypes.bool,
    }),
    item: PropTypes.bool,
    offset: PropTypes.oneOfType([
        PropTypes.number,
        PropTypes.string,
        PropTypes.shape({
            xs: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
            sm: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
            md: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
            lg: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
            xl: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
        }),
    ]),
    width: PropTypes.oneOfType([
        PropTypes.number,
        PropTypes.string,
        PropTypes.shape({
            xs: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
            sm: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
            md: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
            lg: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
            xl: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
        }),
    ]),
};

export default Grid;
