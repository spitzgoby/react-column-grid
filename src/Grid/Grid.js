import { addMissingSizes, sizes } from "../utils/breakpoints";
import classNames from "classnames";
import {
    generateSizeClassNames,
    getAdjustedLayoutProps,
    getGap,
} from "./Grid.layout";
import PropTypes from "prop-types";
import React from "react";
import ThemeContext from "../ThemeContext";

import "./Grid.scss";

const defaultClear = false;
const defaultHide = false;
const defaultOffset = 0;
const defaultWidth = 12;
const shortHandProps = ["offset", "width"];

const isInteger = (value) => !isNaN(parseInt(value, 10));
const useShorthandSyntax = (propName, prop) =>
    shortHandProps.includes(propName) && isInteger(prop);

const Grid = ({
    children,
    className,
    container,
    gap: propGap = null,
    hide = {},
    item,
    offset = {},
    width = { xs: 12 },
}) => {
    // Determine breakpoints and update state
    // State is used to ensure that the context provider does not trigger
    // unnecessary renders
    const { gap } = React.useContext(ThemeContext);

    const shouldUsePropGap = container && propGap;
    const calculatedGap = shouldUsePropGap ? propGap : gap;
    const [adjustedGap, setAdjustedGap] = React.useState(calculatedGap);

    if (shouldUsePropGap && adjustedGap !== propGap) {
        setAdjustedGap(calculatedGap);
    }

    // Fill out incomplete layout props
    const adjustedHide = addMissingSizes(
        "hide",
        hide,
        defaultHide,
        useShorthandSyntax
    );
    const adjustedOffset = addMissingSizes(
        "offset",
        offset,
        defaultOffset,
        useShorthandSyntax
    );
    const adjustedWidth = addMissingSizes(
        "width",
        width,
        defaultWidth,
        useShorthandSyntax
    );

    // Generate classes based on width, offset and sizes
    const sizeClassNames = generateSizeClassNames({
        hide: adjustedHide,
        item,
        offset: adjustedOffset,
        width: adjustedWidth,
    });

    const getThemeValues = () => ({ gap: adjustedGap });

    const getClass = () =>
        classNames({ "rcg-c": container }, sizeClassNames, className);

    const getStyle = () => ({
        gap: container && adjustedGap && getGap(adjustedGap),
    });

    const renderChildren = () => {
        let currentColumns = { xs: 0, sm: 0, md: 0, lg: 0, xl: 0 };
        const childrenToRender = Array.isArray(children)
            ? children
            : [children];

        return childrenToRender.map((child, index) => {
            let renderedChild = child;

            if (child?.type?.name === Grid.name && child?.props?.item) {
                const renderedChildOffset = sizes.reduce((acc, size) => {
                    const completedClear = addMissingSizes(
                        "clear",
                        child.props.clear,
                        defaultClear,
                        useShorthandSyntax
                    );
                    const completedHide = addMissingSizes(
                        "hide",
                        child.props.hide,
                        defaultHide,
                        useShorthandSyntax
                    );
                    const completedOffset = addMissingSizes(
                        "offset",
                        child.props.offset,
                        defaultOffset,
                        useShorthandSyntax
                    );
                    const completedWidth = addMissingSizes(
                        "width",
                        child.props.width,
                        defaultWidth,
                        useShorthandSyntax
                    );
                    const { adjustedColumn, adjustedOffset } =
                        getAdjustedLayoutProps(
                            size,
                            completedOffset,
                            completedWidth,
                            completedHide,
                            completedClear,
                            currentColumns
                        );

                    currentColumns[size] = adjustedColumn;
                    acc[size] = adjustedOffset;

                    return acc;
                }, {});
                const renderedChildProps = {
                    ...child.props,
                    key: index,
                    offset: renderedChildOffset,
                };

                renderedChild = <Grid {...renderedChildProps} />;
            }

            return renderedChild;
        });
    };

    const renderWithThemeProvider = () => (
        <ThemeContext.Provider value={getThemeValues()}>
            {renderGrid()}
        </ThemeContext.Provider>
    );

    const renderGrid = () => (
        <div className={getClass()} style={getStyle()}>
            {container ? renderChildren() : children}
        </div>
    );

    return shouldUsePropGap ? renderWithThemeProvider() : renderGrid();
};

Grid.propTypes = {
    children: PropTypes.node,
    clear: PropTypes.shape({
        xs: PropTypes.bool,
        sm: PropTypes.bool,
        md: PropTypes.bool,
        lg: PropTypes.bool,
        xl: PropTypes.bool,
    }),
    className: PropTypes.string,
    container: PropTypes.bool,
    gap: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    hide: PropTypes.shape({
        xs: PropTypes.bool,
        sm: PropTypes.bool,
        md: PropTypes.bool,
        lg: PropTypes.bool,
        xl: PropTypes.bool,
    }),
    item: PropTypes.bool,
    offset: PropTypes.oneOfType([
        PropTypes.number,
        PropTypes.string,
        PropTypes.shape({
            xs: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
            sm: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
            md: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
            lg: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
            xl: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
        }),
    ]),
    width: PropTypes.oneOfType([
        PropTypes.number,
        PropTypes.string,
        PropTypes.shape({
            xs: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
            sm: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
            md: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
            lg: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
            xl: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
        }),
    ]),
};

export default Grid;
