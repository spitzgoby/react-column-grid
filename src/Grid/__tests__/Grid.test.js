import { render } from "@testing-library/react";
import Grid from "../Grid";
import * as layout from "../Grid.layout";
import React from "react";
import * as breakpoints from "../../utils/breakpoints";

describe("<Grid />", () => {
    beforeAll(() => {
        jest.spyOn(layout, "generateSizeClassNames");
        jest.spyOn(breakpoints, "addMissingSizes");
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    describe("when styling component", () => {
        it("should calculate props for all sizes when creating styles", () => {
            const hideProp = { xs: true, md: false };
            const offsetProp = { xs: "0", md: "2", lg: "3" };
            const widthProp = { xs: "6", md: "4", lg: "3" };

            render(
                <Grid
                    container
                    hide={hideProp}
                    offset={offsetProp}
                    width={widthProp}
                />
            );

            expect(layout.generateSizeClassNames).toHaveBeenCalledWith({
                hide: { xs: true, sm: true, md: false, lg: false, xl: false },
                item: undefined,
                offset: { xs: "0", sm: "0", md: "2", lg: "3", xl: "3" },
                width: { xs: "6", sm: "6", md: "4", lg: "3", xl: "3" },
            });
        });

        it("should use shorthand syntax for width and offset", () => {
            const offsetProp = "1";
            const widthProp = "10";

            render(
                <Grid
                    container
                    hide={{}}
                    offset={offsetProp}
                    width={widthProp}
                />
            );

            expect(layout.generateSizeClassNames).toHaveBeenCalledWith({
                hide: { xs: false, sm: false, md: false, lg: false, xl: false },
                item: undefined,
                offset: { xs: "1", sm: "1", md: "1", lg: "1", xl: "1" },
                width: { xs: "10", sm: "10", md: "10", lg: "10", xl: "10" },
            });
        });

        it("should use shorthand syntax for clear", () => {
            render(
                <Grid container>
                    <Grid item clear />
                </Grid>
            );

            expect(breakpoints.addMissingSizes).toHaveBeenCalledWith(
                "clear",
                true,
                false,
                expect.any(Function)
            );
        });

        it("should use supplied class", () => {
            layout.generateSizeClassNames.mockReturnValueOnce("test-class");

            const { container } = render(
                <Grid item hide={{ xs: true }}></Grid>
            );

            expect(container.querySelector(".test-class")).not.toBeNull();
        });

        it("should use supplied container class if it is a container", () => {
            const { container } = render(<Grid container />);

            expect(container.querySelector(".rcg-c")).not.toBeNull();
        });
    });

    describe("when the grid is a container", () => {
        it("should create a grid container", () => {
            const { container } = render(<Grid container />);

            expect(container).toMatchSnapshot();
        });

        it("should be able to render a single child component", () => {
            const renderGrid = () => {
                render(
                    <Grid container item>
                        <div></div>
                    </Grid>
                );
            };

            expect(renderGrid).not.toThrow();
        });

        it("should not update props of children that are not grid items", () => {
            const { container } = render(
                <Grid container>
                    <div width={4} />
                    <div width={4} />
                </Grid>
            );

            expect(container).toMatchSnapshot();
        });

        it("should increase the offset of child grid items in each size class", () => {
            const { container } = render(
                <Grid container>
                    <Grid
                        item
                        width={{ xs: 12, md: 3 }}
                        offset={{ xs: 0, md: 3 }}
                    />
                    <Grid
                        item
                        width={{ xs: 12, md: 3 }}
                        offset={{ xs: 0, md: 3 }}
                    />
                </Grid>
            );

            expect(container).toMatchSnapshot();
        });

        it("should move elements onto the next row when the columns exceed row width", () => {
            const { container } = render(
                <Grid container>
                    <Grid
                        item
                        width={{ xs: 6, md: 3 }}
                        offset={{ xs: 0, md: 3 }}
                    />
                    <Grid
                        item
                        width={{ xs: 6, md: 3 }}
                        offset={{ xs: 2, md: 3 }}
                    />
                    <Grid
                        item
                        width={{ xs: 12, md: 3 }}
                        offset={{ xs: 0, md: 3 }}
                    />
                </Grid>
            );

            expect(container).toMatchSnapshot();
        });

        it("should recalculate rows when an item spills over into a new row", () => {
            const { container } = render(
                <Grid container>
                    <Grid item width={{ xs: 6 }} offset={{ xs: 3 }} />
                    <Grid item width={{ xs: 8 }} />
                    <Grid item width={{ xs: 4 }} />
                </Grid>
            );

            expect(container).toMatchSnapshot();
        });

        it("should clear move an item onto the nest row if the child clears the row", () => {
            const { container } = render(
                <Grid container>
                    <Grid
                        item
                        width={{ xs: 1 }}
                        clear={{ xs: true, md: false }}
                    />
                    <Grid item width={{ xs: 1 }} />
                    <Grid item width={{ xs: 1 }} />
                </Grid>
            );

            expect(container).toMatchSnapshot();
        });

        it("should update gap state if it has changed", () => {
            const startingGap = 1.5;
            const changedGap = 2;
            const { rerender } = render(<Grid container gap={startingGap} />);
            const setGapSpy = jest.fn();
            const useStateSpy = jest
                .spyOn(React, "useState")
                .mockImplementation(() => [startingGap, setGapSpy]);

            rerender(<Grid container gap={changedGap} />);

            expect(setGapSpy).toHaveBeenCalledWith(changedGap);

            useStateSpy.mockRestore();
        });
    });
});
