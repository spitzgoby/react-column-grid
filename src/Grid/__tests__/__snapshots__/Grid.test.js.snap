// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Grid /> when the grid is a container should clear move an item onto the nest row if the child clears the row 1`] = `
<div>
  <div
    class="rcg-c"
  >
    <div
      class="rcg-xs-1-1 rcg-sm-1-1 rcg-md-1-1 rcg-lg-1-1 rcg-xl-1-1"
    />
    <div
      class="rcg-xs-1-1 rcg-sm-1-1 rcg-md-2-1 rcg-lg-2-1 rcg-xl-2-1"
    />
    <div
      class="rcg-xs-2-1 rcg-sm-2-1 rcg-md-3-1 rcg-lg-3-1 rcg-xl-3-1"
    />
  </div>
</div>
`;

exports[`<Grid /> when the grid is a container should create a grid container 1`] = `
<div>
  <div
    class="rcg-c"
  />
</div>
`;

exports[`<Grid /> when the grid is a container should increase the offset of child grid items in each size class 1`] = `
<div>
  <div
    class="rcg-c"
  >
    <div
      class="rcg-xs-1-12 rcg-sm-1-12 rcg-md-4-3 rcg-lg-4-3 rcg-xl-4-3"
    />
    <div
      class="rcg-xs-1-12 rcg-sm-1-12 rcg-md-10-3 rcg-lg-10-3 rcg-xl-10-3"
    />
  </div>
</div>
`;

exports[`<Grid /> when the grid is a container should move elements onto the next row when the columns exceed row width 1`] = `
<div>
  <div
    class="rcg-c"
  >
    <div
      class="rcg-xs-1-6 rcg-sm-1-6 rcg-md-4-3 rcg-lg-4-3 rcg-xl-4-3"
    />
    <div
      class="rcg-xs-3-6 rcg-sm-3-6 rcg-md-10-3 rcg-lg-10-3 rcg-xl-10-3"
    />
    <div
      class="rcg-xs-1-12 rcg-sm-1-12 rcg-md-4-3 rcg-lg-4-3 rcg-xl-4-3"
    />
  </div>
</div>
`;

exports[`<Grid /> when the grid is a container should not update props of children that are not grid items 1`] = `
<div>
  <div
    class="rcg-c"
  >
    <div
      width="4"
    />
    <div
      width="4"
    />
  </div>
</div>
`;

exports[`<Grid /> when the grid is a container should recalculate rows when an item spills over into a new row 1`] = `
<div>
  <div
    class="rcg-c"
  >
    <div
      class="rcg-xs-4-6 rcg-sm-4-6 rcg-md-4-6 rcg-lg-4-6 rcg-xl-4-6"
    />
    <div
      class="rcg-xs-1-8 rcg-sm-1-8 rcg-md-1-8 rcg-lg-1-8 rcg-xl-1-8"
    />
    <div
      class="rcg-xs-9-4 rcg-sm-9-4 rcg-md-9-4 rcg-lg-9-4 rcg-xl-9-4"
    />
  </div>
</div>
`;
